### PR TITLE
Add Shift+Click to extend terminal text selection (iTerm2-style)

### LINKED ISSUE #3426: Shift+click for text selection doesn't work
### Discord username (optional)

_No response_

### Describe the bug

I usually use Shift+click many times a day for selecting a block of text, it works in every program I know of except Warp. it works for the text that I can type for the last block, but it doesn't work for previous blocks. Please fix it

### To reproduce

On a block that is not the most recent one, select a character with a mouse, release the click, press and hold Shift on keyboard then move your cursor on a few characters or lines futher and click once.

### Expected behavior

It should the select the text between the original position where you first clicked or selected some text and the wherer you clicked last

### Screenshots

_No response_

### Operating system

MacOS

### Operating system and version

MacOS 12.5.1

### Shell Version

_No response_

### Current Warp version

v0.2023.07.18.08.03.stable_00

### Regression

No, this bug or issue has existed throughout my experience using Warp

### Recent working Warp date

not sure, I'm new

### Additional context

_No response_

### Does this block you from using Warp daily?

Yes, this issue prevents me from using Warp daily.

### Is this a Warp specific issue? (i.e. does it happen in Terminal, iTerm, Kitty, etc.)

Yes, this I confirmed this only happens in Warp, not other terminals.

### Warp Internal (ignore): linear-label:b8107fdf-ba31-488d-b103-d271c89cac3e

None

### PR DESCRIPTION
**Pull request by cola-runner**
## Summary

Adds iTerm2-style **Shift+Click to extend a terminal text selection**. With a selection already present, **Shift + left-click** moves the nearest selection boundary to the clicked point:

- clicking before the selection extends the start/head side
- clicking after the selection extends the end/tail side
- clicking inside the selection shrinks whichever side is closer

A plain click still starts a new selection, and Shift+Click+drag can continue fine-tuning the extended text selection. This makes selecting long / multi-screen output much easier: pick a start, scroll, then Shift+Click the other side instead of drag-and-auto-scrolling the whole way.

Closes #3426

(Supersedes the duplicate reports #11932, #9963, and #4715 -- same iTerm2-style Shift+Click extend behavior.)

## How it works

The shared selection model now exposes nearest-boundary extension logic, so block-list and alt-screen selections use the same behavior.

- **Selection model** (`selection.rs`): adds reusable nearest-boundary logic for moving start vs end when extending. The moved boundary remains the active update endpoint so Shift+Click+drag keeps fine-tuning the same side.
- **Block list** (`block_list_element.rs`, `view.rs`): Shift+Click dispatches `Extend` when a simple text selection already exists and the click is not on a link / revealed secret. Plain drag still updates the drag endpoint, while Shift+Click+drag continues dispatching text-selection updates instead of being misclassified as block selection.
- **Alt screen** (`alt_screen_element.rs`, `alt_screen.rs`): Shift+Click extension is wired for full-screen apps as well.

## Scope / follow-ups

- AI rich-content block internal selections are intentionally left out of scope for this PR because they are implemented differently.

## Testing

Verified after rebasing onto latest `upstream/master` (`c20645b9`) with this PR applied:

- `cargo fmt`
- `cargo test -p warp nearest_boundary --lib` -- 13 passed
- `cargo test -p warp test_alt_screen_extend_selection_without_existing_selection_noops --lib` -- 1 passed
- `cargo check -p warp`

The `nearest_boundary` coverage includes model tests plus view-level regressions for keeping plain drag updates on the drag endpoint and keeping Shift+Click+drag on the boundary selected by Extend.

Earlier macOS GUI evidence for the cross-screen block-list workflow is here: https://gist.github.com/cola-runner/37a80d8cf984abe661964e1b9d3fa9a9

Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

**Comment by cola-runner** [comment]
@cla-bot check

**Comment by cola-runner** [comment]
Updated the Testing section with macOS GUI verification for the cross-screen workflow, and linked the reviewed evidence:

https://gist.github.com/cola-runner/37a80d8cf984abe661964e1b9d3fa9a9

/oz-review

File: app/src/terminal/view.rs [hunk 1]
```diff
@@ -18854,6 +18866,41 @@ impl TerminalView {
         ctx.notify();
     }
 
+    /// Extends the existing block-text selection to `point`, keeping the
+    /// original anchor (head) fixed and moving only the tail. This backs
```
Comment (seemeroland) [inline app/src/terminal/view.rs]: I don't think this is the correct behavior, looking at what iTerm2 does. It may move either the head or the tail. If the shift-clicked point is before the head, it extends the head.

Also if the shift-clicked point is within the selection, it may update the head if the point is closer to the head (shrink the head vs shrink the tail, depending which side the point is closer to). Using the row is probably a fine approximation.

File: app/Cargo.toml [hunk 2]
```diff
@@ -837,6 +837,7 @@ traces = []
 viewing_shared_sessions = []
 workflow_aliases = []
 rect_selection = []
+shift_click_extend_selection = []
```
Comment (seemeroland) [inline app/Cargo.toml]: nit: it should be okay to not add the feature flag. We're going to have to enable it in all channels anyway

File: app/src/terminal/view.rs [hunk 3]
```diff
@@ -17988,6 +17988,13 @@ impl TerminalView {
             SelectAction::Update {
                 point, side, delta, ..
             } => self.update_alt_selection(*point, *side, delta, ctx),
+            SelectAction::Extend { point, side, .. } => {
+                // Shift+click extension is only wired up for the block list
+                // today; for the alt screen we treat it as a plain tail update
+                // so the variant is handled exhaustively and stays correct if
+                // alt-screen extension is enabled later.
+                self.update_alt_selection(*point, *side, &Lines::zero(), ctx);
```
Comment (seemeroland) [inline app/src/terminal/view.rs]: It'd be great to add alt screen support as well (e.g. in vim). They should be able to share some of the logic. 

We'd probably want to add this feature to AIBlocks as well but that is implemented more differently so no need to worry about it in this PR.

File: app/src/terminal/view.rs [hunk 4]
```diff
@@ -18854,6 +18866,41 @@ impl TerminalView {
...
+    fn extend_block_text_selection(
+        &mut self,
+        point: BlockListPoint,
+        side: Side,
+        position: Vector2F,
+        ctx: &mut ViewContext<Self>,
+    ) {
+        let has_selection = self.model.lock().block_list().selection().is_some();
+        if !has_selection {
+            self.begin_block_text_selection(point, side, SelectionType::Simple, position, ctx);
```
Comment (seemeroland) [inline app/src/terminal/view.rs]: Probably worth a comment that this is unexpected because extending only makes sense with a selection, and this is a fallback to treat it like a regular click. Alternatively I think it might be fine to do nothing in this case

**Review by seemeroland — COMMENTED** [review]
Thanks for taking a stab at this!

**Comment by cola-runner** [comment]
Updated this PR on top of latest `master` (`a30cc7a3`) and addressed the review feedback:

- changed Shift+Click extension to move the nearest selection boundary, matching iTerm2 behavior for extending before/after the selection and shrinking from inside it
- shared the nearest-boundary logic between block-list and alt-screen selections
- wired Shift+Click extension for alt screen selections
- removed the compile-time feature flag from the final diff
- changed the no-selection `Extend` fallback to no-op with a comment, since Extend should only be dispatched when a selection exists
- kept Shift+Click+drag as a text-selection update path so the extended selection can still be fine-tuned while Shift is held

Validation on latest `master` plus this PR:

- `cargo test -p warp nearest_boundary --lib` (6 passed)
- `cargo check -p warp`

**Comment by cola-runner** [comment]
Added another small test-only commit for edge coverage:

- same-row shrinking uses column distance to pick the nearest boundary
- reversed block-list selections update the correct stored anchor/tail
- block-list and alt-screen `Extend` no-op when no selection exists

Validation after the test commit:

- `cargo test -p warp nearest_boundary --lib` (10 passed)
- `cargo test -p warp test_alt_screen_extend_selection_without_existing_selection_noops --lib` (1 passed)
- `cargo check -p warp`

**Comment by cola-runner** [comment]
Hi @seemeroland, I addressed your review feedback in the latest updates:

- Shift+Click now moves the nearest selection boundary, including shrinking from inside the current selection.
- Block-list and alt-screen selections share the nearest-boundary logic.
- Alt-screen Shift+Click extension is wired.
- The compile-time feature flag was removed.
- `Extend` without an existing selection now no-ops.
- Shift+Click+drag can continue fine-tuning the extended text selection while Shift is held.
- Added edge tests for same-row shrink behavior, reversed block-list selections, and no-selection no-op behavior.

Validation:

- `cargo test -p warp nearest_boundary --lib` (10 passed)
- `cargo test -p warp test_alt_screen_extend_selection_without_existing_selection_noops --lib` (1 passed)
- `cargo check -p warp`
- GitHub CI is green on the current head (`352260410bda3d935c2e7c295555657e12923d20`).

Would you mind taking another look when you have a chance?

**Comment by cola-runner** [comment]
Hi @seemeroland, following up here. I rebased this PR onto latest `master` (`f554cf76`) and re-ran the local validation:

- `cargo test -p warp nearest_boundary --lib` (10 passed)
- `cargo test -p warp test_alt_screen_extend_selection_without_existing_selection_noops --lib` (1 passed)
- `cargo check -p warp`

All previous review threads are resolved, and the current head is `8ccfbbecc97f3ac907b78b45de5d0853148b3644`.

cc @harryalbert @adrawbond-alto @BobbyWang0120 since you have recently touched nearby terminal view / mouse dispatch / terminal event paths. Would one of you be able to take a look or help route this to the right owner?

**Comment by cola-runner** [comment]
Hi @seemeroland, following up after this PR went stale and became conflicting.

I rebased it onto latest `master` (`c20645b9`) and pushed the updated head `5b6e4f43d453eb4a3dff99f781879dfec1e00d13`. GitHub now reports the PR as mergeable again.

Local validation after the rebase:

- `cargo fmt`
- `cargo test -p warp nearest_boundary --lib` (10 passed)
- `cargo test -p warp test_alt_screen_extend_selection_without_existing_selection_noops --lib` (1 passed)
- `cargo check -p warp`

cc @harryalbert @adrawbond-alto @BobbyWang0120 in case one of you can help review or route this to the right owner.

**Comment by cola-runner** [comment]
Hi @seemeroland, one more update before review: I did another self-review pass and found a boundary-wiring issue around Shift+Click+drag.

Fixed in `2eba005927ef47209e597d2abd220bc3a60e7f48`:

- plain block-list drag updates now keep using the drag endpoint instead of re-running nearest-boundary selection
- Shift+Click Extend now moves the nearest boundary and keeps that moved boundary as the active update endpoint, so Shift+Click+drag continues fine-tuning the same side
- added model/view regression coverage for the block-list and alt-screen cases

Validation after the fix:

- `cargo fmt`
- `cargo test -p warp nearest_boundary --lib` (13 passed)
- `cargo test -p warp test_alt_screen_extend_selection_without_existing_selection_noops --lib` (1 passed)
- `cargo check -p warp`
